### PR TITLE
PythonScopeJpyImpl: Disable Conversion Cache

### LINKED ISSUE #5112: The auto conversion feature of the Python UDF decorator has a performance problem.
This was found by the latest bench marking effort to measure the performance impact from the usability improvement on Python UDF.

### PR DESCRIPTION
**Pull request by nbauernfeind**
This fixes a memory leak discovered in #5112 via repeated runs of:

```python
import time, jpy
from deephaven import empty_table, garbage_collect

row_count = 10_000_000

begin_time = time.perf_counter_ns()
source = empty_table(row_count).update(["X = repeat(ii % 250, 100)"]) 
empty_table(1).select('Y = 1')
print('Rows / Sec:', row_count / ((time.perf_counter_ns() - begin_time) / 1_000_000_000))

Runtime = jpy.get_type('java.lang.Runtime')
print('Gigs Used Before GC:',
        (Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()) / 1024 / 1024 / 1024)

del source;
del row_count;
del begin_time
for i in range(10):
    garbage_collect()

print('Gigs Used After GC:',
        (Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()) / 1024 / 1024 / 1024)
del Runtime
```

Nightlies running here: https://github.com/nbauernfeind/deephaven-core/actions/runs/7833860264/

### CODE REVIEW

File: engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java [hunk 1]
```diff
@@ -189,14 +190,43 @@ public final class QueryLanguageParser extends GenericVisitorAdapter<Class<?>, Q
...
+     * {@link #getResult()}} method.
+     *
+     * @param expression The query language expression to parse
+     * @param packageImports Wildcard package imports
+     * @param classImports Individual class imports
+     * @param staticImports Wildcard static imports. All static variables and methods for the given classes are
+     *        imported.
+     * @param variables A map of the names of scope variables to their types
+     * @param variableTypeArguments A map of the names of scope variables to their type arguments
+     * @param possibleParams A mutable map of the names of query scope variables to their values
```
Comment (jmao-denver) [inline engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java:217]: Don't quite like the name `possibleParams`, I think `queryScopeVariables` might be easier for readers.

**Review by jmao-denver — DISMISSED** [review]
1. weak-values cache makes sense for the main global scope
2. like the decoupling of QueryScope from QLP

File: engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java (hunk 1 above)
Comment (rcaudy) [reply engine/table/src/main/java/io/deephaven/engine/table/impl/lang/QueryLanguageParser.java:217]: This is a good point, and I endorse it whole-heartedly.

**Review by rcaudy — COMMENTED** [review]
.

File: engine/context/src/main/java/io/deephaven/engine/context/StandaloneQueryScope.java [hunk 2]
```diff
@@ -77,9 +79,13 @@ public <T> void putParam(final String name, final T value) {
     }
 
     @Override
-    public Map<String, Object> toMap() {
-        return valueRetrievers.entrySet().stream()
-                .collect(Collectors.toMap(Map.Entry::getKey, e -> e.getValue().value));
+    public Map<String, Object> toMap(@NotNull final Predicate<Map.Entry<String, Object>> predicate) {
+        final HashMap<String, Object> result = new HashMap<>();
+        valueRetrievers.entrySet().stream()
+                .map(e -> Map.entry(e.getKey(), (Object) e.getValue().value))
```
[the remark this answers is not in the document]
Comment (nbauernfeind) [reply engine/context/src/main/java/io/deephaven/engine/context/StandaloneQueryScope.java]: as discussed offline, this is to unwrap the value retriever. I assume the retriever is allowed to return null, but that the retrievers themselves are non-null.

File: engine/table/src/main/java/io/deephaven/engine/table/impl/select/AbstractConditionFilter.java [hunk 3]
```diff
@@ -98,8 +94,9 @@ public synchronized void init(TableDefinition tableDefinition) {
 
         try {
             final QueryScope queryScope = ExecutionContext.getContext().getQueryScope();
-            final Map<String, Object> possibleParams = queryScope.toMap();
-            for (Map.Entry<String, Object> param : possibleParams.entrySet()) {
+            final Map<String, Object> queryScopeVariables = queryScope.toMap(
```
[the remark this answers is not in the document]
Comment (nbauernfeind) [reply engine/table/src/main/java/io/deephaven/engine/table/impl/select/AbstractConditionFilter.java:97]: I'm willing to do this if this is your preference. I was considering the other usages. They both filter to Table instances -- unfortunately they also need to unwrap first, and probably we would prefer to unwrap once.

File: engine/table/src/main/java/io/deephaven/engine/util/PythonEvaluatorJpy.java [hunk 4]
```diff
@@ -47,6 +47,8 @@ public static PythonEvaluatorJpy withGlobals() throws IOException, InterruptedEx
 
     private PythonEvaluatorJpy(PyDictWrapper globals) {
         this.globals = globals;
+        // forcefully load this helper class which imports python modules
+        PyCallableWrapperJpyImpl.init();
```
Comment (rcaudy) [inline engine/table/src/main/java/io/deephaven/engine/util/PythonEvaluatorJpy.java:51]: I'm worried this is some kind of band-aid allowing us to mask a deeper issue.

File: server/src/main/java/io/deephaven/server/console/python/PythonImportInitializer.java [hunk 5]
```diff
@@ -0,0 +1,17 @@
+package io.deephaven.server.console.python;
+
+import io.deephaven.engine.util.PyCallableWrapperJpyImpl;
+import io.deephaven.integrations.python.PythonObjectWrapper;
+
+public abstract class PythonImportInitializer {
```
Comment (rcaudy) [inline server/src/main/java/io/deephaven/server/console/python/PythonImportInitializer.java:6]: Feels like this pattern could be replaced with service loader.

File: engine/table/src/main/java/io/deephaven/engine/util/PythonScope.java [hunk 6]
```diff
@@ -160,14 +160,8 @@ default Collection<String> getKeysCollection() {
      */
     default Map<String, Object> getEntriesMap() {
```
Comment (rcaudy) [inline engine/table/src/main/java/io/deephaven/engine/util/PythonScope.java:161]: This method feels inconsistent. Do we still need it?

File: engine/table/src/main/java/io/deephaven/engine/util/PythonScope.java [hunk 7]
```diff
@@ -134,13 +135,12 @@ default Stream<String> getKeys() {
...
-     * and the values via {@link #convertValue(Object)}.
+     * but the values still need conversion via {@link #convertValue(Object)}.
      *
-     * @return the string keys and converted values
+     * @return the string keys and original values
      */
-    default Stream<Entry<String, Object>> getEntries() {
-        return getEntriesRaw()
-                .map(e -> new SimpleImmutableEntry<>(convertStringKey(e.getKey()), convertValue(e.getValue())));
+    default Stream<Entry<String, PyObj>> getEntries() {
```
Comment (rcaudy) [inline engine/table/src/main/java/io/deephaven/engine/util/PythonScope.java:142]: To your point, this adds very little value now.

File: server/src/main/java/io/deephaven/server/console/python/PythonImportInitializer.java (hunk 5 above)
Comment (devinrsmith) [reply server/src/main/java/io/deephaven/server/console/python/PythonImportInitializer.java:6]: I think the problem w/ service loader is that the classes are on the classpath regardless of whether we are running python or groovy. In this current pattern, it's only invoked when we are coming from known-python script session.

**Review by rcaudy — APPROVED** [review]

**Pull request merged**